### ISSUE TITLE
unrestrictedTraverse breaks on Zope 4 / Python 2.7 for unicode strings/paths

### ISSUE THREAD

**Issue by jugmac00** [zopefoundation/Zope#715]
I have code like the following... which works with Zope 2 / Python 2.7
```
        ret = {}
        dbChooser = self.unrestrictedTraverse('/Control_Panel/Database')
        dbNames = list(dbChooser.getDatabaseNames())
        if 'temporary' in dbNames:
            dbNames.remove('temporary')
        for dbName in dbNames:
            path = '/Control_Panel/Database/%s' % dbName
            dbMan = self.unrestrictedTraverse(path)
```

In Zope 4 / Python 2 it breaks, `getDatabaseNames` now returns a list of unicode strings and `unrestrictedTraverse` does a check whether the path is a "str" - if not it converts the unicode path into a list... as follows...

`[u'/', u'C', u'o', u'n', u't', u'r', u'o', u'l', u'_', u'P', u'a', u'n', u'e', u'l', u'/', u'D', u'a', u't', u'a', u'b', u'a', u's', u'e', u'/', u'b', u'l', u'i', u's', u's', u'_', u'd', u'a', u't', u'a']`

**Related code**:
https://github.com/zopefoundation/Zope/blob/01487e5daf3ba4d8add1c12bfb46e4acaf27cac0/src/OFS/Traversable.py#L174-L178

btw there was quite some discussion in #674 - but without a solution (as the problem was more focussed on Python 3, and not Python 2).

P.S.: my local fix looks like...
```
        #if isinstance(path, str):
        import six
        if isinstance(path, six.string_types):
```

If this sounds reasonable I'll create a pr to fix this issue on the Zope 4 branch.

**Comment by dataflake** [comment]
As explained in #718, your own code needs fixing. You're breaking expected semantics by trying to fix it inside `unrestrictedTraverse`.

**Comment by jugmac00** [comment]
@dataflake Thank you for your feedback.

Unfortunately, I do not concur with your conclusion.

Let me explain why.

The code I maintain connects two framework functions (`unrestrictedTraverse` and `getDatabaseNames`).

**Behaviour**
- Zope 2 and Python 2 works
- Zope 4 and Python 2 breaks
- Zope 4 and Python 3 works (presumably)

Without looking at the code this smells like the problem of the framework not the application's code.
(Whether the problem is within `unrestrictedTraverse` or `getDatabaseNames` is another topic).

I think it should not be the duty of an application developer to do some heavy lifting and unicode/byte encoding/decoding when using two framework functions which both handle "paths".

> I would also consider your issue #715 ad invalid because you're asking for something that was explicitly not supported, see the comment inside unrestrictedTraverse.

The comment is 11+ years old and predates the arrival of Zope 4 - with which unicode was allowed for object ids - and thus paths and urls, see #166 and #181 

I never made up my mind about whether this is valid for Python 2 or 3 before, as I will never use this, but in my current Zope 4 application I am able to create unicode object ids... but I could not access them via `unrestrictedTraverse` - this seems odd.

Concerning the comment - I still think the comment is outdated but at very least for Python 3 the comment is wrong - as for Python 3 unicode is required whereas the comment say "unicode paths are not allowed".

Assuming I am dead wrong with my assumptions, I see another problem with the existing code. I was able to change the code from checking against `str` to `text_types` without any test breaking.

https://github.com/zopefoundation/Zope/blob/01487e5daf3ba4d8add1c12bfb46e4acaf27cac0/src/OFS/Traversable.py#L174-L178

Ideally, there would be a strong test harness which enforces the "contract".

The current code disallows unicode strings for Python 2, but allows and works with a list of unicode strings.

**tl/dr**
This very few lines of code led to quite some confusion in #674 and for me too, so at very least - instead of closing this issue -  I'd like the comment to be fixed, maybe a proper docstring created and ideally a test written which shows that the contract is enforced.

**Comment by ale-rt** [comment]
@jugmac00 I also think this issue should be closed.

I would patch the code you showed us to convert `dbName` into a native string, something like:
```
from Products.CMFPlone.utils import safe_nativestring
path = '/Control_Panel/Database/%s' % safe_nativestring(dbName)
```

I know that it might be pretty easy to support both cases, but it has been like this for years and I would not spend time changing it now :)

**Comment by jugmac00** [comment]
@ale-rt Thank you for taking your time.

Well, I think I have to bow to the majority :-)

Thank you very much for not ignoring to update the comment ( via #720 ), as I think it is important to keep our code base up to date an as unambiguous as possible - especially for the second or third generation Zope developers.

And anyhow - the "problem" only exists for the combination of Zope 4 and Python 2 which I hopefully do not have to maintain for too long.

**Comment by ale-rt** [comment]
Actually a similar issue will occur also on Zope 4 and Python 3 if somebody will pass bytes (e.g. using `.encode`).

**Comment by jugmac00** [comment]
Accidentially closed - as I assumed the pull requests by @ale-rt were already merged.

**Comment by dataflake** [comment]
The comment change PRs #720 and #721 are now merged, so this can be closed.
